[onert] Introduce experimental training APIs

hseok-oh opened the pull request:
This commit introduces experimental training APIs.

ONE-DCO-1.0-Signed-off-by: Hyeongseok Oh <hseok82.oh@samsung.com>

hseok-oh commented:
Two different style

- Add APIs on `nnfw_session` session
  - Split APIs as possible
- Introduce APIs on new `nnfw_train_session` session
  - Merge APIs as possible

glistening commented on runtime/onert/api/include/nnfw_experimental_train.h:
```diff
@@ -0,0 +1,182 @@
+ * @param[out]  count   The number of training model inputs
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_get_input_count(nnfw_train_session *session, uint32_t *count);
```
@hseok-oh We may need use different name to distinguish the same name for inference (`nnfw_session` series).

hseok-oh commented:
Updated

- Remove `nnfw_train_session`
- Split APIs to set input & expected
- Remove `nnfw_set_train_output`
- Add index to get loss for multiple output
- Add `update_weights` argument for validate mode

jyoungyun commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,193 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ *                      If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_set_train_expected(nnfw_session *session, uint32_t index, void *expected,
```
When will this function be used? Is this function used with `nnfw_run` api ?

jyoungyun commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,193 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+  /** Learning rate */
+  float learning_rate = 0.001f;
+  /** Batch size */
+  uint32_t batch_size = 1;
```
TODO
- loss
- optimizer

hseok-oh commented:
@nnfw-bot test onert-x64-release

glistening commented:
@hseok-oh If you determined not to introduce `nnfw_training_session`, what about using `nnfw_train` for training related APIs? For example,

- `nnfw_prepare_train` → `nnfw_tᅟrain_prepare`
- `nnfw_set_train_input` → `nnfw_train_set_input`
- `nnfw_set_train_expected` → `nnfw_train_set_expected`
- `nnfw_train` → `nnfw_train` ( ? )
- `nnfw_get_loss` → `nnfw_train_get_loss`
- `nnfw_export_inference_model` → `nnfw_train_export_model`
- `nnfw_get_train_input_count` → `nnfw_train_get_input_count`
...

hseok-oh commented:
I've updated function name. PTAL

glistening commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @param[in] path    The path to export inference model
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_export_inference_model(nnfw_session *session, const char *path);
```
Will we export other type of model? I don't have an idea about other type of models except for inference model. Maybe we can export checkpoint later — `nnfw_train_export_checkpoint`.

glistening commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @param[out]  count   The number of training model inputs
+ * @return @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_input_count(nnfw_session *session, uint32_t *count);
```
What is the difference between model input and train input? I guessed the count would be same.

glistening commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @param[in]   session The session to get the number of training model expected outputs
+ * @param[out]  count   The number of training model expected outputs
+ */
+NNFW_STATUS nnfw_train_expected_count(nnfw_session *session, uint32_t *count);
```
Also, I guessed expected count == model output count. Why do we need this API?

hseok-oh replied:
Let's remove this and `nnfw_train_input_count`.

glistening commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @param session The session to convert training mode to inference mode
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_convert_to_inference(nnfw_session *session);
```
According to @param, it is for setting to training mode. 

```suggestion
NNFW_STATUS nnfw_train_set_training_mode(nnfw_session *session, bool is_training);
```

glistening commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @param session The session to convert inference mode to training mode
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_convert_to_train(nnfw_session *session);
```
We may remove this API by `nnfw_train_set_training_mode(sess, false)`.

jyoungyun commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,194 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+typedef struct nnfw_train_info
+{
+  /** Learning rate */
+  float learning_rate = 0.001f;
```
How about separating the train preparation info and execution info?
The `nnfw_train_prepare` does not need learning_rate information. This is needed in `nnfw_train`.

After the trained graph is generated once, the user can use this graph with various learning rate and epoch. Do we need a scenario like this?

hseok-oh replied:
>How about separating the train preparation info and execution info?

I didn't consider this yet. But we may need this later. 
I supposed that we will get all parameter for training before `prepare` as our first step.

>After the trained graph is generated once, the user can use this graph with various learning rate

`nnfw_train_set_traininfo` API may be used for this.

hseok-oh commented:
I've updated
- Remove in/expected count APIs
- Rename export circle API
- Comment out non-essential APIs
- Fix typo

jyoungyun reviewed (dismissed):
API design looks good to me. :)

I made a draft PR using this changes. #10900

ragmani commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
+                                 nnfw_tensorinfo *input_info);
```
Is it okay not to pass the size of buffer?

hseok-oh replied:
`input_info` will be used to get size.

ragmani commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ *                        If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
```
```suggestion
NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const void *input,
```

ragmani commented on runtime/onert/api/src/nnfw_api.cc:
```diff
@@ -385,3 +385,53 @@ NNFW_STATUS nnfw_pop_pipeline_output(nnfw_session *session, void *outputs)
+  return session->train_prepare(info);
+}
+
+NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, void *input,
```
```suggestion
NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const void *input,
```

ragmani commented on runtime/onert/api/include/nnfw_experimental.h:
```diff
@@ -152,4 +156,168 @@ NNFW_STATUS nnfw_push_pipeline_input(nnfw_session *session, void *inputs, void *
+ *                      If it is nullptr, it will not change shape and batch size
+ * @return  @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, void *expected,
```
```suggestion
NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, const void *expected,
```

ragmani commented on runtime/onert/api/src/nnfw_api.cc:
```diff
@@ -385,3 +385,53 @@ NNFW_STATUS nnfw_pop_pipeline_output(nnfw_session *session, void *outputs)
+  return session->train_set_input(index, input, input_info);
+}
+
+NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, void *expected,
```
```suggestion
NNFW_STATUS nnfw_train_set_expected(nnfw_session *session, uint32_t index, const void *expected,
```

ragmani commented on runtime/onert/api/src/nnfw_api_internal.cc:
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
   return NNFW_STATUS_NO_ERROR;
 }
+
+NNFW_STATUS nnfw_session::train_prepare(nnfw_train_info *info)
```
```suggestion
NNFW_STATUS nnfw_session::train_prepare(const nnfw_train_info *info)
```

ragmani commented on runtime/onert/api/src/nnfw_api_internal.cc:
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
+  return NNFW_STATUS_ERROR;
+}
+
+NNFW_STATUS nnfw_session::train_set_input(uint32_t index, void *input,
```
```suggestion
NNFW_STATUS nnfw_session::train_set_input(uint32_t index, const void *input,
```

ragmani commented on runtime/onert/api/src/nnfw_api_internal.cc:
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
+  return NNFW_STATUS_ERROR;
+}
+
+NNFW_STATUS nnfw_session::train_set_expected(uint32_t index, void *expected,
```
```suggestion
NNFW_STATUS nnfw_session::train_set_expected(uint32_t index, const void *expected,
```

ragmani reviewed (commented):
API design looks good to me. :)

glistening reviewed (approved):
LGTM

jyoungyun reviewed (dismissed):
LGTM

hseok-oh commented:
I've updated to add APIs to get tensorinfo. PTAL

jyoungyun commented on runtime/onert/api/src/nnfw_api_internal.h:
```diff
@@ -163,6 +163,8 @@ struct nnfw_session
   NNFW_STATUS set_backends_per_operation(const char *backend_settings);
 
   NNFW_STATUS train_prepare(const nnfw_train_info *info);
+  NNFW_STATUS train_input_tensorinfo(uint32_t index, nnfw_tensorinfo *ti);
+  NNFW_STATUS train_expected_tensorinfo(uint32_t index, nnfw_tensorinfo *ti);
```
👍

jyoungyun commented on runtime/onert/api/src/nnfw_api_internal.cc:
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
+  }
+
+  // Use default info if info is null, otherwise use given info
+  (void)info;
```
If the info is null, how about treating it as an error to clarify operation?

hseok-oh replied:
It is designed to use training info in model metadata if `info` param is `nullptr`. Please refer API comment in `nnfw_experimental.h`.

jyoungyun reviewed (approved):
LGTM

glistening reviewed (approved):
LGTM

hseok-oh commented on runtime/onert/api/src/nnfw_api_internal.cc:
```diff
@@ -1129,3 +1129,159 @@ NNFW_STATUS nnfw_session::set_backends_per_operation(const char *backend_setting
+    std::cerr << "Error during nnfw_session::train_set_input : input buffer is null" << std::endl;
+    return NNFW_STATUS_UNEXPECTED_NULL;
+  }
+
```
It is designed to use shape in model and batch training info if `input_tensorinfo` is `nullptr`. Please refer API comment in `nnfw_experimental.h`
